### PR TITLE
MHCADDI model

### PR DESCRIPTION
**Pull request by sebastiandro**
# Summary
 
_Please provide a high-level summary of the changes for the changes and notes for the reviewers_
 
- [ ] Unit tests provided for these changes
- [ ] Documentation and docstrings added for these changes using the [sphinx style](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html)

## Changes 

* Implementation of the [MHCADDI model](https://arxiv.org/pdf/1905.00534.pdf) based on their [original code](https://github.com/andreeadeac22/graph_coattention).

### CODE REVIEW

**Comment by benedekrozemberczki** [comment]
@sebastiandro I will look into the segmentation indices.

File: chemicalx/models/mhcaddi.py [hunk 1]
```diff
@@ -1,17 +1,478 @@
...
+
+    def message_aggregation(self, node, msg, seg_i):
+        msg = torch.zeros_like(node).index_add(0, seg_i, msg)
+        return msg
+
+
+class CoAttention(nn.Module):
+    """The co-attention network for MHCADDI model."""
+
+    def __init__(self, d_in: int, d_out: int, n_head: int = 1, dropout: float = 0.1):
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:96]: please improve the variable names through this PR - prefixing variable names with `d_` and `n_` is very confusing

Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py:96]: @cthoyt I will overtake and refactor this.

File: chemicalx/models/mhcaddi.py [hunk 2]
```diff
@@ -1,17 +1,502 @@
...
+
+
+def segment_softmax(
+    logit: torch.FloatTensor,
+    n_seg: torch.LongTensor,
+    seg_i: torch.LongTensor,
+    idx_j: torch.LongTensor,
+    temperature: torch.FloatTensor,
+):
+    """Segmentation softmax calculation."""
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:39]: please add parameter comments

File: chemicalx/models/mhcaddi.py [hunk 3]
```diff
@@ -1,17 +1,502 @@
...
+    return seg_i
+
+
+class MessagePassing(nn.Module):
+    """A network for creating node representations based on internal message passing."""
+
+    def __init__(self, d_node: int, d_edge: int, d_hid: int, dropout=0.1):
+        """Instantiate the MessagePassing network.
+
+        :param d_node: Dimension of node features
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:60]: please improve variable names

Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py:60]: I will.

File: chemicalx/models/mhcaddi.py (hunk 2 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py:39]: I will.

**Comment by benedekrozemberczki** [comment]
Will not merge without a complete rewrite of the variable names and docstrings.

File: chemicalx/models/mhcaddi.py [hunk 4]
```diff
@@ -1,17 +1,502 @@
...
+        seg_g2,
+        node2,
+        edge2,
+        inn_seg_i2,
+        inn_idx_j2,
+        out_seg_i2,
+        out_idx_j2,
+    ):
+        """Make a forward pass with the data."""
+        for step_i in range(self.n_prop_step):
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:251]: use zip over self.mps, self.coats, and self.lns

File: chemicalx/models/mhcaddi.py [hunk 5]
```diff
@@ -1,17 +1,502 @@
...
+        out_idx_j = [
+            list(range(0, right_graph_size)) * left_graph_size
+            for left_graph_size, right_graph_size in zip(graph_sizes_left, graph_sizes_right)
+        ]
+        out_idx_j = functools.reduce(operator.iconcat, out_idx_j, [])
+        out_idx_j = torch.tensor(out_idx_j) + shift_sums_right
+
+        return out_seg_i, out_idx_j
+
+    def unpack(self, batch: DrugPairBatch):
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:455]: this functionality should be inside forward()

**Comment by cthoyt** [comment]
After that's all done I'll take a other pass at review

**Comment by benedekrozemberczki** [comment]
@cthoyt Simplified it massively.

**Review by benedekrozemberczki — APPROVED** [review]

**Comment by benedekrozemberczki** [comment]
@cthoyt  can I merge this?

File: chemicalx/models/mhcaddi.py [hunk 6]
```diff
@@ -1,17 +1,507 @@
...
+
+        node_left_center = self.key_projection(node_left).index_select(0, segmentation_index_left)
+        node_right_center = self.key_projection(node_right).index_select(0, segmentation_index_right)
+
+        node_left_neighbor = self.value_projection(node_right).index_select(0, segmentation_index_right)
+        node_right_neighbor = self.value_projection(node_left).index_select(0, segmentation_index_left)
+
+        translation = (node_left_center * node_right_center).sum(1)
+
+        # TODO factor out a helper function that can be applied to both sides to reduce code
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:152]: TODO

File: chemicalx/models/mhcaddi.py [hunk 7]
```diff
@@ -1,17 +1,507 @@
...
+
+        self.co_attention = CoAttention(
+            input_channels=hidden_channels,
+            output_channels=hidden_channels,
+            dropout=dropout,
+        )
+
+        self.linear = nn.LayerNorm(hidden_channels)
+
+        # FIXME why is nn.LeakyReLU used inside init of Linear()
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:210]: FIXME

File: chemicalx/models/mhcaddi.py [hunk 8]
```diff
@@ -1,17 +1,507 @@
...
+        self.prediction_readout_projection = nn.Linear(hidden_channels, readout_channels, nn.LeakyReLU())
+
+    def update_fn(self, atom_features: torch.Tensor, inner_message: torch.Tensor, outer_message: torch.Tensor):
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:213]: Necessary? would be more transparent not to have this function

File: chemicalx/models/mhcaddi.py [hunk 9]
```diff
@@ -1,17 +1,507 @@
...
+        :param outer_segmentation_index_right: Heads of edges connecting atoms between right and left drug molecules
+        :param outer_index_right: Heads of edges connecting atoms between right and left drug molecules
+        :returns: Graph level representations.
+        """
+        inner_message_left = self.message_passing(atom_left, bond_left, inner_segmentation_index_left, inner_index_left)
+        inner_message_right = self.message_passing(
+            atom_right, bond_right, inner_segmentation_index_right, inner_index_right
+        )
+
+        # TODO can this come before the calculation of inner messages? If so, it should
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:257]: todo

File: chemicalx/models/mhcaddi.py [hunk 10]
```diff
@@ -1,17 +1,507 @@
...
+        outer_message_left, outer_message_right = self.co_attention(
+            atom_left,
+            outer_segmentation_index_left,
+            outer_index_left,
+            atom_right,
+            outer_segmentation_index_right,
+            outer_index_right,
+        )
+
+        # TODO make helper function that can de-duplicate code
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:267]: TODO

File: chemicalx/models/mhcaddi.py [hunk 11]
```diff
@@ -1,17 +1,507 @@
...
+        :param segmentation_molecule_right:  Mapping from node id to graph id for the right drugs.
+        :param atom_type_right: Atom types of the atoms in the right drug molecules.
+        :param atom_features_right: Features of the atoms in the right drug molecules..
+        :param bond_type_right: Bond types in the right drug molecules.
+        :param inner_segmentation_index_right: Heads of edges connecting atoms within the right drug molecules.
+        :param inner_index_right: Tails of edges connecting atoms within the right drug molecules.
+        :param graph_sizes_right: Graph size vector on the right.
+        :returns: A column vector of predicted scores.
+        """
+        # TODO factor out duplicate code that can be applied independently to left and right sides
```
Comment (cthoyt) [inline chemicalx/models/mhcaddi.py:380]: TODO

File: chemicalx/utils.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
+"""Utilities for chemicalx."""
```
Comment (cthoyt) [inline chemicalx/utils.py:1]: please provide unit tests for these helper functions

File: examples/mhcaddi_examples.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
+"""Example with DeepSynergy."""
```
Comment (cthoyt) [inline examples/mhcaddi_examples.py:1]: fix docs... this doesn't seem very carefully done

File: examples/mhcaddi_examples.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+from chemicalx import pipeline
+from chemicalx.data.datasetloader import TwoSides
+from chemicalx.models.mhcaddi import MHCADDI
+
+
+def main():
+    """Train and evaluate the DeepSynergy model."""
+    dataset = TwoSides()
+    model = MHCADDI(atom_feature_channels=69, atom_type_channels=100, bond_type_channels=12)
```
Comment (cthoyt) [inline examples/mhcaddi_examples.py:11]: arbitrary constants? should be coming from dataset

File: tests/unit/test_models.py [hunk 15]
```diff
@@ -188,8 +188,18 @@ def test_gcnbmp(self):
 
     def test_mhcaddi(self):
         """Test MHCADDI."""
-        model = MHCADDI(x=2)
-        assert model.x == 2
+        model = MHCADDI(atom_feature_channels=69, atom_type_channels=100, bond_type_channels=12)
```
Comment (cthoyt) [inline tests/unit/test_models.py:191]: arbitrary constants?

**Review by cthoyt — CHANGES_REQUESTED** [review]
Please address the TODOs in the code highlighted in the PR

File: chemicalx/models/mhcaddi.py (hunk 7 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py]: Done.

File: chemicalx/models/mhcaddi.py (hunk 9 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py]: Done.

File: chemicalx/models/mhcaddi.py (hunk 11 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py]: Done.

File: examples/mhcaddi_examples.py (hunk 13 above)
Comment (benedekrozemberczki) [reply examples/mhcaddi_examples.py]: Fixed.

File: chemicalx/utils.py (hunk 12 above)
Comment (benedekrozemberczki) [reply chemicalx/utils.py:1]: Provided tests through the softmax segment - that reuses both.

File: chemicalx/models/mhcaddi.py (hunk 10 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py]: Done.

File: chemicalx/models/mhcaddi.py (hunk 6 above)
Comment (benedekrozemberczki) [reply chemicalx/models/mhcaddi.py]: Refactored.
